File: Workbench/util/TimeUtil.py
from datetime import datetime, timezone , timedelta

def print_now():
    """Prints the current local datetime."""
    print(datetime.now())

def print_now_utc():
    """Prints the current UTC datetime."""
    print(datetime.now(timezone.utc))

def get_now():
    """Returns the current local datetime."""
    return datetime.now()

def get_now_utc():
    """Returns the current UTC datetime."""
    return datetime.now(timezone.utc)

def get_timestamp():
    """Returns the current timestamp in seconds."""
    return datetime.now().timestamp()

def get_now_utc_string():
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

def get_now_hkt_string():
    return datetime.now(timezone.utc).astimezone(timezone(timedelta(hours=8))).strftime("%Y-%m-%d %H:%M:%S")

def get_now_utc_date():
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")

def get_utc_date(backward_days: int = 0):
    """
    Returns the current UTC date minus the specified number of days.
    :param backward_days: Number of days to subtract from the current date.
    :return: Current UTC date minus the specified number of days.
    """
    return (datetime.now(timezone.utc) - timedelta(days=backward_days)).strftime("%Y-%m-%d")

def get_utc_now_ms():
    """
    Returns the current UTC time in milliseconds.
    :return: Current UTC time in milliseconds.
    """
    return int(datetime.now(timezone.utc).timestamp() * 1000)

def get_hkt_now_ms():
    """
    Returns the current Hong Kong Time (HKT) in milliseconds.
    :return: Current HKT time in milliseconds.
    """
    return int(datetime.now(timezone.utc).astimezone(timezone(timedelta(hours=8))).timestamp() * 1000)

def get_latency_ms(event_ts: int) -> int:
    return event_ts - get_utc_now_ms()

File: cpp/TimeUtil.cpp
#include "TimeUtil.h"
#include <iostream>
#include <iomanip>

namespace TimeUtil {

using namespace std::chrono;

void printNow() {
    auto now = system_clock::now();
    std::time_t tt = system_clock::to_time_t(now);
    std::cout << std::put_time(std::localtime(&tt), "%Y-%m-%d %H:%M:%S") << std::endl;
}

void printNowUTC() {
    auto now = system_clock::now();
    std::time_t tt = system_clock::to_time_t(now);
    std::cout << std::put_time(std::gmtime(&tt), "%Y-%m-%d %H:%M:%S") << std::endl;
}

system_clock::time_point getNow() {
    return system_clock::now();
}

system_clock::time_point getNowUTC() {
    return system_clock::now();
}

double getTimestamp() {
    return duration<double>(system_clock::now().time_since_epoch()).count();
}

std::string getNowUTCString() {
    auto now = system_clock::now();
    std::time_t tt = system_clock::to_time_t(now);
    char buf[20];
    std::strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", std::gmtime(&tt));
    return buf;
}

std::string getNowHKTString() {
    auto now = system_clock::now();
    std::time_t tt = system_clock::to_time_t(now + hours(8));
    char buf[20];
    std::strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", std::gmtime(&tt));
    return buf;
}

std::string getNowUTCDate() {
    auto now = system_clock::now();
    std::time_t tt = system_clock::to_time_t(now);
    char buf[11];
    std::strftime(buf, sizeof(buf), "%Y-%m-%d", std::gmtime(&tt));
    return buf;
}

std::string getUTCDate(int backwardDays) {
    auto now = system_clock::now() - hours(24 * backwardDays);
    std::time_t tt = system_clock::to_time_t(now);
    char buf[11];
    std::strftime(buf, sizeof(buf), "%Y-%m-%d", std::gmtime(&tt));
    return buf;
}

long long getUTCNowMs() {
    return duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
}

long long getHKTNowMs() {
    return duration_cast<milliseconds>((system_clock::now() + hours(8)).time_since_epoch()).count();
}

long long getLatencyMs(long long eventTs) {
    return eventTs - getUTCNowMs();
}

long long getUTCTsNs() {
    return duration_cast<nanoseconds>(system_clock::now().time_since_epoch()).count();
}

long long msToNs(long long ms) {
    return ms * 1'000'000;
}

} // namespace TimeUtil

